alf/algorithms/predictive_representation_learner.py

```diff
@@ -347,34 +409,47 @@ def _multi_step_latent_rollout(self, init_latent, num_unroll_steps,
     def train_step(self, exp: TimeStep, state):
         # [B, num_unroll_steps + 1]
         info = exp.rollout_info
+        targets = common.as_list(info.target)
         batch_size = exp.step_type.shape[0]
         latent, state = self._encoding_net(exp.observation, state)
 
         sim_latent = self._multi_step_latent_rollout(
             latent, self._num_unroll_steps, info.action, state)
 
-        # [num_unroll_steps + 1)*B, ...]
-        train_info = self._decoder.train_step(sim_latent).info
-        train_info_spec = dist_utils.extract_spec(train_info)
-        train_info = dist_utils.distributions_to_params(train_info)
-        train_info = alf.nest.map_structure(
-            lambda x: x.reshape(self._num_unroll_steps + 1, batch_size, *x.
-                                shape[1:]), train_info)
-        # [num_unroll_steps + 1, B, ...]
-        train_info = dist_utils.params_to_distributions(
-            train_info, train_info_spec)
-        target = alf.nest.map_structure(lambda x: x.transpose(0, 1),
-                                        info.target)
-        loss_info = self._decoder.calc_loss(target, train_info, info.mask.t())
-        loss_info = alf.nest.map_structure(lambda x: x.mean(dim=0), loss_info)
+        loss = 0
+        for i, decoder in enumerate(self._decoders):
+            # [num_unroll_steps + 1)*B, ...]
+            train_info = decoder.train_step(sim_latent).info
+            train_info_spec = dist_utils.extract_spec(train_info)
+            train_info = dist_utils.distributions_to_params(train_info)
+            train_info = alf.nest.map_structure(
+                lambda x: x.reshape(self._num_unroll_steps + 1, batch_size, *x.
+                                    shape[1:]), train_info)
+            # [num_unroll_steps + 1, B, ...]
+            train_info = dist_utils.params_to_distributions(
+                train_info, train_info_spec)
+            target = alf.nest.map_structure(lambda x: x.transpose(0, 1),
+                                            targets[i])
+            loss_info = decoder.calc_loss(target, train_info, info.mask.t())
+            loss_info = alf.nest.map_structure(lambda x: x.mean(dim=0),
+                                               loss_info)
+            loss += loss_info.loss
+
+        loss_info = LossInfo(loss=loss, extra=loss)
 
         return AlgStep(output=latent, state=state, info=loss_info)
 
     @torch.no_grad()
     def preprocess_experience(self, experience: Experience):
         """Fill experience.rollout_info with PredictiveRepresentationLearnerInfo
 
-        Note that the shape of experience is [B, T, ...]
+        Note that the shape of experience is [B, T, ...].
+
+        The target is a Tensor with the shape of [B, T, unroll_steps+1, ...]
+        when there is only one decoder. When multiple decorders are used,
```

Comment:
This is not very accurate. Even with one decode, the target could be a nest because a field can correspond to a nest.

Reply:
Updated the comments.